File: labs/control-ui/api-shadow/camouflage_stream_masquerader.ts
export type MasqueradeProbeVerdict = 'ACCEPT_STREAM' | 'DEFLECT_TO_DECOY' | 'DROP_CONNECTION';

function areBytesEqual(a: Uint8Array, b: Uint8Array): boolean {
  if (a.length !== b.length) return false;
  for (let i = 0; i < a.length; i++) {
    if (a[i] !== b[i]) return false;
  }
  return true;
}

export class CamouflageStreamMasquerader {
  private validUserIds: Uint8Array[] = [];
  private sharedSecret: Uint8Array;
  public decoyHost: string;
  constructor(sharedSecret: Uint8Array, decoyHost: string) {
    this.sharedSecret = sharedSecret;
    this.decoyHost = decoyHost;
  }

  registerUser(userId: Uint8Array): void {
    const copy = new Uint8Array(16);
    copy.set(userId.subarray(0, 16));
    this.validUserIds.push(copy);
  }

  generatePreamble(userId: Uint8Array): Uint8Array {
    const out = new Uint8Array(32);
    const salt = new Uint8Array(16);
    salt.fill(0x5a);
    out.set(salt, 0);

    const secLen = this.sharedSecret.length > 0 ? this.sharedSecret.length : 1;
    for (let i = 0; i < 16; i++) {
      const secByte = this.sharedSecret.length > 0 ? this.sharedSecret[i % secLen] : 0;
      const mask = secByte! ^ salt[i]!;
      const uByte = i < userId.length ? userId[i] : 0;
      out[16 + i] = uByte! ^ mask;
    }
    return out;
  }

  inspectInboundStream(preamble: Uint8Array): MasqueradeProbeVerdict {
    if (preamble.length < 32) return 'DEFLECT_TO_DECOY';

    const salt = preamble.subarray(0, 16);
    const candidateUid = new Uint8Array(16);
    const secLen = this.sharedSecret.length > 0 ? this.sharedSecret.length : 1;

    for (let i = 0; i < 16; i++) {
      const secByte = this.sharedSecret.length > 0 ? this.sharedSecret[i % secLen] : 0;
      const mask = secByte! ^ salt[i]!;
      candidateUid[i] = preamble[16 + i]! ^ mask;
    }

    for (const valid of this.validUserIds) {
      if (areBytesEqual(valid, candidateUid)) {
        return 'ACCEPT_STREAM';
      }
    }
    return 'DEFLECT_TO_DECOY';
  }
}

File: labs/mobile/donor-archive/tunnel/CamouflageStreamMasquerader.kt
package com.luminet.android.tunnel

enum class AndroidProbeAction {
    ACCEPT_STREAM, DEFLECT_TO_DECOY, DROP_CONNECTION
}

class CamouflageStreamMasquerader(
    private val sharedSecret: ByteArray,
    val decoyHost: String
) {
    private val validUserIds = mutableListOf<ByteArray>()

    fun registerUser(userId: ByteArray) {
        validUserIds.add(userId.copyOf(16))
    }

    fun generatePreamble(userId: ByteArray): ByteArray {
        val out = ByteArray(32)
        val salt = ByteArray(16) { 0x5a.toByte() }
        System.arraycopy(salt, 0, out, 0, 16)

        val secLen = if (sharedSecret.isNotEmpty()) sharedSecret.size else 1
        for (i in 0 until 16) {
            val secByte = if (sharedSecret.isNotEmpty()) sharedSecret[i % secLen] else 0
            val mask = (secByte.toInt() xor salt[i].toInt()).toByte()
            val uByte = if (i < userId.size) userId[i] else 0
            out[16 + i] = (uByte.toInt() xor mask.toInt()).toByte()
        }
        return out
    }

    fun inspectInboundStream(preamble: ByteArray): AndroidProbeAction {
        if (preamble.size < 32) return AndroidProbeAction.DEFLECT_TO_DECOY

        val salt = preamble.copyOfRange(0, 16)
        val candidateUid = ByteArray(16)
        val secLen = if (sharedSecret.isNotEmpty()) sharedSecret.size else 1

        for (i in 0 until 16) {
            val secByte = if (sharedSecret.isNotEmpty()) sharedSecret[i % secLen] else 0
            val mask = (secByte.toInt() xor salt[i].toInt()).toByte()
            candidateUid[i] = (preamble[16 + i].toInt() xor mask.toInt()).toByte()
        }

        for (valid in validUserIds) {
            if (valid.contentEquals(candidateUid)) {
                return AndroidProbeAction.ACCEPT_STREAM
            }
        }
        return AndroidProbeAction.DEFLECT_TO_DECOY
    }
}
